testsuite/smoke-test/src/consensus/consensus_fault_tolerance.rs

```diff
@@ -0,0 +1,432 @@
+        let mut small_rng = SmallRng::from_entropy();
+
+        let now = Instant::now();
+        let mut index = 0;
+        while !finish.load(Ordering::Relaxed) {
+            let sender = small_rng.gen_range(0usize, accounts.len() - 1);
+            let (a, b) = accounts.split_at_mut(sender + 1);
+            transfer_coins_non_blocking(
+                &validator_clients[small_rng.gen_range(0usize, validator_clients.len())].1,
+                &transaction_factory,
+                &mut a[sender],
+                &b[small_rng.gen_range(0, b.len())],
+                1,
+            )
+            .await;
+
+            index += 1;
+
+            let elapsed = now.elapsed().as_secs_f32();
+            let wanted = (1 + index) as f32 / tps;
+            if elapsed < wanted {
+                tokio::time::sleep(Duration::from_secs_f32(wanted - elapsed)).await;
+            } else if elapsed > wanted + 1.0 {
+                info!("Traffic is running {}s behind", elapsed - wanted);
+            }
+        }
+    });
+    ActiveTrafficGuard {
+        finish_traffic: finish_copy,
+    }
+}
+
+async fn run_fail_point_test(
+    num_validators: usize,
+    cycles: usize,
+    cycle_duration_s: f32,
+    parts_in_cycle: usize,
+    traffic_tps: f32,
+    max_block_size: u64,
+    // (cycle, part) -> (Vec(validator_index, name, action), reset_old_enpoints)
+    get_fail_points_to_set: Box<
+        dyn FnMut(usize, usize) -> (Vec<(usize, String, String)>, bool) + Send,
+    >,
+    // (cycle, executed_epochs, executed_rounds, executed_transactions, current_state, previous_state)
+    check_cycle: Box<dyn FnMut(usize, u64, u64, u64, Vec<NodeState>, Vec<NodeState>)>,
+) {
+    let mut swarm = create_swarm(num_validators, max_block_size).await;
+    let _active_traffic = if traffic_tps > 0.0 {
+        start_traffic(5, traffic_tps, &mut swarm).await
+    } else {
+        ActiveTrafficGuard {
+            finish_traffic: Arc::new(AtomicBool::new(false)),
+        }
+    };
+    test_consensus_fault_tolerance(
+        &mut swarm,
+        cycles,
+        cycle_duration_s,
+        parts_in_cycle,
+        Box::new(FailPointFailureInjection::new(get_fail_points_to_set)),
+        check_cycle,
+        false,
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn test_no_failures() {
+    let num_validators = 3;
+
+    let mut swarm = create_swarm(num_validators, 1).await;
+
+    test_consensus_fault_tolerance(
+        &mut swarm,
+        3,
+        5.0,
+        1,
+        no_failure_injection(),
+        Box::new(move |_, _, executed_rounds, executed_transactions, _, _| {
+            assert!(
+                executed_transactions >= 4,
+                "no progress with active consensus, only {} transactions",
+                executed_transactions
+            );
+            assert!(
+                executed_rounds >= 2,
+                "no progress with active consensus, only {} rounds",
+                executed_rounds
+            );
```

Comment:
How do you get the assert values? Is this based on cycle duration and round timeout?